tests/test_redcap_sync.py

```diff
@@ -196,3 +221,182 @@ def test_collapse_housing_fields():
+    matching_ids = ["1111", "3333"]
+    matching_records, unmatching_records = select_records_by_cdms_matches(
+        input_df,
+        input_redcap_records,
+        matched_external_ids=matching_ids,
+        external_id_col=external_id_col,
+    )
+    pd.testing.assert_frame_equal(matching_records, expected_matching_df)
+    pd.testing.assert_frame_equal(unmatching_records, expected_not_matching_df)
+
+
+def test_add_reject_status_columns(patch_datetime_now):
+    dob_field = "dob"
+    external_id_col = "cdms_id"
+    input_df = pd.DataFrame({"record_id": ["1", "2", "3"]}, index=[1, 2, 3])
+    timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+    reason = f"mismatched {dob_field} and {external_id_col}"
+    expected_output_df = pd.DataFrame(
+        {
+            "record_id": ["1", "2", "3"],
+            "integration_status": [REDCAP_REJECTED_PERSON for i in range(3)],
+            "integration_status_timestamp": [timestamp for i in range(3)],
+            "integration_status_reason": [reason for i in range(3)],
+        },
+        index=[1, 2, 3],
+    )
+    output_df = add_integration_status_columns(
+        input_df, status=REDCAP_REJECTED_PERSON, reason=reason
+    )
+    pd.testing.assert_frame_equal(expected_output_df, output_df)
+
+
+def test_handle_cdms_matching(patch_datetime_now):
+    input_df = pd.DataFrame(
+        {
+            "record_id": ["1", "2", "3"],
+            "cdms_id": ["1111", "2222", "3333"],
+            "external_id": ["1111", "2222", "3333"],
+            "dob": [None, "1953-03-17", "1933-02-04"],
+            "other_stuff": ["some", "more", "values"],
+        },
+        index=[1, 2, 3],
+    )
+    input_redcap_records = pd.DataFrame(
+        {
+            "record_id": ["1", "2", "3"],
+            "cdms_id": ["1111", "2222", "3333"],
+            "dob": [None, "1953-03-17", "1933-02-04"],
+            "other_stuff": ["some", "more", "values"],
+        },
+        index=[1, 2, 3],
+    )
+
+    # Expect handle_cdms_matching to return the matching records.
+    expected_matched_cdms_ids = ["2222"]
+    expected_output_df = pd.DataFrame(
+        {
+            "record_id": ["2"],
+            "cdms_id": ["2222"],
+            "external_id": ["2222"],
+            "dob": ["1953-03-17"],
+            "other_stuff": ["more"],
+        },
+        index=[2],
+    )
+
+    # Expect import_records_to_redcap to be called with the following data.
+    expected_error_status_columns = {
+        REDCAP_INTEGRATION_STATUS: [REDCAP_REJECTED_PERSON for i in range(2)],
+        REDCAP_INTEGRATION_STATUS_TIMESTAMP: [
+            datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S") for i in range(2)
+        ],
+        REDCAP_INTEGRATION_STATUS_REASON: [
+            "mismatched dob and cdms_id" for i in range(2)
+        ],
+    }
+    expected_unmatched_records = pd.DataFrame(
+        {"record_id": ["1", "3"], **expected_error_status_columns}, index=[1, 3]
+    )
+
+    with patch(
```

Comment:
Nitpick: This is a little hard for me to follow (not that I know of a better way).